File: node/deliberation/test/guardrail.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { enforceGuardrails } from '../src/guardrail.mjs';

test('guardrail: no hard blocks + no soft conflicts → passes clean', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [],
      proposed_position: { estimated_dollar: 20, agent_position_dollar_floor: 10 },
    },
    { verdict: 'buy' },
  );
  assert.equal(r.passed, true);
  assert.equal(r.hard_blocks.length, 0);
  assert.equal(r.soft_conflicts.length, 0);
  assert.equal(r.policy_applied, 'no_conflicts');
});

test('guardrail: hard block → passed=false', () => {
  const r = enforceGuardrails(
    { gate_state: { blocking: ['capital_halt'] }, accretions: [], proposed_position: {} },
    { verdict: 'buy' },
  );
  assert.equal(r.passed, false);
  assert.deepEqual(r.hard_blocks, ['capital_halt']);
});

test('guardrail: unconfirmed accretion opposing → soft conflict', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [
        {
          id: 99,
          operator_confirmed: false,
          canonical_path: 'per-symbol:BTC-USDT',
          fact: 'PAUSE BUYS during high VIX windows',
        },
      ],
      proposed_position: { estimated_dollar: 15, agent_position_dollar_floor: 10 },
    },
    { verdict: 'buy' },
  );
  assert.equal(r.soft_conflicts.length, 1);
  // floor=10, threshold=max(20, 25)=25, proposed=15 → flag
  assert.equal(r.policy_applied, 'flag');
  assert.equal(r.exposure_threshold, 25);
});

test('guardrail: soft conflict + above threshold → pause', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [
        { id: 99, operator_confirmed: false, fact: 'NO BUY in this window', canonical_path: 'x' },
      ],
      proposed_position: { estimated_dollar: 60, agent_position_dollar_floor: 20 },
    },
    { verdict: 'buy' },
  );
  // floor=20, threshold=max(40, 25)=40, proposed=60 → pause
  assert.equal(r.policy_applied, 'pause');
  assert.equal(r.exposure_threshold, 40);
});

test('guardrail: confirmed accretion does NOT trigger soft conflict', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [
        { id: 99, operator_confirmed: true, fact: 'pause buys', canonical_path: 'x' },
      ],
      proposed_position: { estimated_dollar: 100, agent_position_dollar_floor: 10 },
    },
    { verdict: 'buy' },
  );
  assert.equal(r.soft_conflicts.length, 0);
  assert.equal(r.policy_applied, 'no_conflicts');
});

test('guardrail: minimum absolute pause floor is $25', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [{ id: 1, operator_confirmed: false, fact: 'no buy', canonical_path: 'x' }],
      proposed_position: { estimated_dollar: 10, agent_position_dollar_floor: 5 },
    },
    { verdict: 'buy' },
  );
  // floor=5, threshold=max(10, 25)=25, proposed=10 → flag (below 25)
  assert.equal(r.exposure_threshold, 25);
  assert.equal(r.policy_applied, 'flag');
});

test('guardrail: hold verdict has no directional soft conflicts', () => {
  const r = enforceGuardrails(
    {
      gate_state: { blocking: [] },
      accretions: [
        { id: 1, operator_confirmed: false, fact: 'no buy under conditions', canonical_path: 'x' },
      ],
      proposed_position: { estimated_dollar: 100, agent_position_dollar_floor: 10 },
    },
    { verdict: 'hold' },
  );
  assert.equal(r.soft_conflicts.length, 0);
});

File: go/deliberation/pkg/deliberation/guardrail_test.go
package deliberation

import (
	"testing"
)

func TestGuardrailCleanPass(t *testing.T) {
	r := EnforceGuardrails(EvidencePackage{
		GateState: GateState{Blocking: []string{}},
		ProposedPosition: &ProposedPosition{
			EstimatedDollar:          20,
			AgentPositionDollarFloor: 10,
		},
	}, Adjudication{Verdict: VerdictBuy})

	if !r.Passed {
		t.Error("expected passed")
	}
	if r.PolicyApplied != "no_conflicts" {
		t.Errorf("policy: %s", r.PolicyApplied)
	}
}

func TestGuardrailHardBlock(t *testing.T) {
	r := EnforceGuardrails(EvidencePackage{
		GateState: GateState{Blocking: []string{"capital_halt"}},
	}, Adjudication{Verdict: VerdictBuy})

	if r.Passed {
		t.Error("expected !passed")
	}
	if len(r.HardBlocks) != 1 || r.HardBlocks[0] != "capital_halt" {
		t.Errorf("hard_blocks: %v", r.HardBlocks)
	}
}

func TestGuardrailSoftConflictAboveThreshold(t *testing.T) {
	r := EnforceGuardrails(EvidencePackage{
		GateState: GateState{Blocking: []string{}},
		Accretions: []Accretion{
			{ID: 99, OperatorConfirmed: false, Fact: "NO BUY in this window", CanonicalPath: "x"},
		},
		ProposedPosition: &ProposedPosition{
			EstimatedDollar:          60,
			AgentPositionDollarFloor: 20,
		},
	}, Adjudication{Verdict: VerdictBuy})

	if r.PolicyApplied != "pause" {
		t.Errorf("policy expected pause, got %s", r.PolicyApplied)
	}
	if r.ExposureThreshold != 40 {
		t.Errorf("threshold expected 40, got %f", r.ExposureThreshold)
	}
}

func TestGuardrailSoftConflictBelowThreshold(t *testing.T) {
	r := EnforceGuardrails(EvidencePackage{
		Accretions: []Accretion{
			{ID: 1, OperatorConfirmed: false, Fact: "no buy", CanonicalPath: "x"},
		},
		ProposedPosition: &ProposedPosition{
			EstimatedDollar:          10,
			AgentPositionDollarFloor: 5,
		},
	}, Adjudication{Verdict: VerdictBuy})

	// floor=5, threshold=max(10, 25)=25, proposed=10 → flag
	if r.ExposureThreshold != 25 {
		t.Errorf("threshold expected 25 (absolute floor), got %f", r.ExposureThreshold)
	}
	if r.PolicyApplied != "flag" {
		t.Errorf("policy expected flag, got %s", r.PolicyApplied)
	}
}

func TestGuardrailConfirmedAccretionIgnored(t *testing.T) {
	r := EnforceGuardrails(EvidencePackage{
		Accretions: []Accretion{
			{ID: 1, OperatorConfirmed: true, Fact: "pause buys", CanonicalPath: "x"},
		},
		ProposedPosition: &ProposedPosition{
			EstimatedDollar:          100,
			AgentPositionDollarFloor: 10,
		},
	}, Adjudication{Verdict: VerdictBuy})

	if len(r.SoftConflicts) != 0 {
		t.Errorf("expected 0 soft conflicts, got %d", len(r.SoftConflicts))
	}
}

func TestFinalVerdictFor(t *testing.T) {
	if FinalVerdictFor(GuardrailResult{Passed: false, HardBlocks: []string{"x"}}, Adjudication{Verdict: VerdictBuy}) != FinalAbort {
		t.Error("hard block should abort")
	}
	if FinalVerdictFor(GuardrailResult{Passed: true, PolicyApplied: "pause"}, Adjudication{Verdict: VerdictBuy}) != FinalOperatorRequired {
		t.Error("pause should require operator")
	}
	if FinalVerdictFor(GuardrailResult{Passed: true, PolicyApplied: "no_conflicts"}, Adjudication{Verdict: VerdictHold}) != FinalHold {
		t.Error("hold verdict should be FinalHold")
	}
	if FinalVerdictFor(GuardrailResult{Passed: true, PolicyApplied: "flag"}, Adjudication{Verdict: VerdictBuy}) != FinalExecute {
		t.Error("flag should still execute")
	}
}
